README.md
# macOS Unlocker V4 for VMware ESXi

## Unlocker 2007-2023
This project is now archived.

The unlocker should continue to run as there have been few changes to the VMware code in many years.
I have stopped developemnt as I no longer use VMware but would be happy to refer to a fork if someone 
sends me an email with the relevant details.

There is also [Auto Unlocker](https://github.com/paolo-projects/auto-unlocker) which is still active.

## 1. Introduction
Unlocker 4 is designed for VMware ESXi 7.

The Unlocker enables certain flags and data tables that are required to see the macOS type when setting
the guest OS type, and modify the implmentation of the virtual SMC controller device. These capabiltiites are normally 
exposed in Fusion and ESXi when running on Apple hardware.

The patch code carries out the following modifications dependent on the product being patched:

* Patch vmx and derivatives to allow macOS to boot
* Patch libvmkctl.so to allow vCenter to boot macOS guests

It is important to understand that the Unlocker cannot add any new capabilities to VMware ESXi
but enables support for macOS that is disabled in the VMware products when run on non-Apple hardware.

The Unlocker cannot:

* add support for new versions of macOS
* add paravirtualized Apple GPU support 
* add older (non-Ryzen) AMD CPU support

or any other features that are not already in the VMware compiled code. 

## 2. Running the patcher

#### The ESXi unlocker will need to be run each time the ESXi Server is upgraded. 
#### It is also best to switch ESXi to Maintanence mode and make sure you do not have any VMs running.

The code is written in Python and has no pre-requisites and should run directly from the release zip download.

* Download a binary release from https://github.com/DrDonk/esxi-unlocker/releases
* Unload the archive to a folder on the ESXi server datastore and extract the files
* Navigate to the folder with the extracted files

You will then need to run one of the following commands to patch or unpatch the ESXi software.

* unlock - apply patches to VMware ESXi
* relock - remove patches from VMware ESXi
* check  - check the patch status of your VMware installation

## 3. FAQS

### 3.1 Remove older versions of ESXi unlocker
If you hve previously used an earlier version of the ESXi Unlocker please uninstall it by logging into the 
support console and running:

```BootModuleConfig.sh --verbose --remove=unlocker.tgz```

and then rebooting the ESXi server.

### 3.2 AMD CPUs
The ESXi Unlocker does not add support for AMD CPUs, that has to be done via settings for the macOS guest runnning on 
a modern AMD CPU. These settings may allow macOS to run based on tests reported back to the project, but there is no 
formal support for AMD CPUs with the unlocker.

A patched macOS AMD kernel or OpenCore must be used to run on older AMD systems.

1. Read this KB article to learn how to edit a guest's VMX file safely https://kb.vmware.com/s/article/2057902
2. Add the following lines were to the VMX file:
```
cpuid.0.eax = "0000:0000:0000:0000:0000:0000:0000:1011"
cpuid.0.ebx = "0111:0101:0110:1110:0110:0101:0100:0111"
cpuid.0.ecx = "0110:1100:0110:0101:0111:0100:0110:1110"
cpuid.0.edx = "0100:1001:0110:0101:0110:1110:0110:1001"
cpuid.1.eax = "0000:0000:0000:0001:0000:0110:0111:0001"
cpuid.1.ebx = "0000:0010:0000:0001:0000:1000:0000:0000"
cpuid.1.ecx = "1000:0010:1001:1000:0010:0010:0000:0011"
cpuid.1.edx = "0000:0111:1000:1011:1111:1011:1111:1111"
vhv.enable = "FALSE"
vpmc.enable = "FALSE"
vvtd.enable = "FALSE"
```
3. Make sure there are no duplicate lines in the VMX file or the guest will not start and a dictionary error will
   be displayed by VMware.
4. You can now install and run macOS as a guest.

## 4. Thanks
Thanks to Zenith432 for originally building the C++ Unlocker and Mac Son of Knife
(MSoK) for all the testing and support.

Thanks also to Sam B for finding the solution for ESXi 6 and helping me with
debugging expertise. Sam also wrote the code for patching ESXi ELF files and
modified the Unlocker code to run on Python 3 in the ESXi 6.5 environment.

Thanks to lucaskamp for testing the new version 4 of ESXi Unlocker.


CHANGELOG.md
# Changes

All dates are UK DD/MM/YY format.

## 03/03/23 4.0.6
### This is the last release of ESXi Unlocker.

_drdonk:_
* Tidy up of code
* Add explanantion that the unlocker cannot add AMD support but some settings in VMX file may work

## 09/01/23 4.0.5
_drdonk:_
* Fixed permissions error correctly this time!

## 03/01/23 4.0.4
_drdonk:_
* `check` command correctly reports if system not patched
* `unlock` command checks for free disk space before patching
* `unlock`  command checks for previous V3 installation before patching
* Fixed error if libvmkctl.so already patched
* Removed an unused unlocker flag (KPPW/KPST) to match Go version
* Update all copyright dates to 2023
* Make sure the files to be patched have write permissions before patching as NFS datastores correctly enforce 
R/W permissions but VMFS3 does not and can patch with read only flag set.

## 22/10/22 4.0.3
_drdonk:_
* Reinstate libvmkctl patch to allow vCenter to boot macOS VMs on ESXi host
* Reinstate boot time load so the libvmkctl patch is loaded when hostd starts
* Store ESXi version with patched files
* `check` command compares current and stored ESXi versions

## 22/09/22 4.0.2
Thanks to _lucaskamp_ and an _anonymous_ tester for testing.
 
_drdonk:_
* Revert to binary file read/writes in patchsmc instead of using mmap, which caused silent errors during the patching 
process leading to failed patches.
* Modified VMTAR format from PSIGNED-XZ to GZIP
* Commands are now required to be run each boot of ESXi to avoid possible "Purple Screens of Death" (PSOD).
* Ensure vmx files have correct permissions, 4555/-r-sr-xr-x,  in apple.v00 archive

## 26/01/22 4.0.1
_drdonk:_
* Fix missing +x bit on checksmc, dumpsmc and relock

## 03/08/22 4.0.0
* Initial release


esx/dumpsmc.py
#!/usr/bin/env python3
# coding=utf-8

# SPDX-FileCopyrightText: © 2014-2021 David Parsons
# SPDX-License-Identifier: MIT

"""
vSMC Header Structure
=====================
Offset  Length  Struct Type Description
----------------------------------------
0x00/00 0x08/08 Q      ptr  Offset to key table
0x08/08 0x04/4  I      int  Number of private keys
0x0C/12 0x04/4  I      int  Number of public keys

vSMC Key Data Structure
Offset  Length  Struct Type Description
----------------------------------------
0x00/00 0x04/04 4s     int  Key name (byte reversed e.g. #KEY is YEK#)
0x04/04 0x01/01 B      byte Length of returned data
0x05/05 0x04/04 4s     int  Data type (byte reversed e.g. ui32 is 23iu)
0x09/09 0x01/01 B      byte Flag R/W
0x0A/10 0x06/06 6x     byte Padding
0x10/16 0x08/08 Q      ptr  Internal VMware routine
0x18/24 0x30/48 48B    byte Data

The internal VMware routines point to 4 variants:
AppleSMCHandleDefault
AppleSMCHandleNTOK
AppleSMCHandleNumKeys
AppleSMCHandleOSK
"""

import mmap
import os.path
import struct
import sys

# Constants for header and key access
HDR_PACK = '=QII'
HDR_LENGTH = 16
KEY_PACK = '=4sB4sB6xQ'
KEY_LENGTH = 24
DATA_LENGTH = 48
ROW_LENGTH = KEY_LENGTH + DATA_LENGTH

# Setup hex string for vSMC headers
# These are the private and public key counts
SMC_HEADER_V0 = b'\xF2\x00\x00\x00\xF0\x00\x00\x00'
SMC_HEADER_V1 = b'\xB4\x01\x00\x00\xB0\x01\x00\x00'

# Keys we use
KEY_KEY = b'YEK#'
LKS_KEY = b'SKL+'
OSK0_KEY = b'0KSO'
OSK1_KEY = b'1KSO'

if sys.version_info < (3, 6):
    sys.stderr.write('You need Python 3.6 or later\n')
    sys.exit(1)


def bytetohex(data):
    return ''.join('{:02X}'.format(c) for c in data).lower()


def printhdr(offset, hdr):
    print(f'File Offset : 0x{offset:08x}')
    print(f'Keys Offset : 0x{hdr[0]:08x}')
    print(f'Private Keys: 0x{hdr[1]:04x}/{hdr[1]}')
    print(f'Public Keys : 0x{hdr[2]:04x}/{hdr[2]}')
    return


def printkey(offset, smc_key, smc_data):
    # Format smc_type as cannot use \ in f-strings
    smc_type = smc_key[2][::-1].replace(b'\x00', b' ').decode('UTF-8')
    print(f'0x{offset:08x} '
          f'{smc_key[0][::-1].decode("UTF-8")} '
          f'{smc_key[1]:02d}  '
          f'{smc_type} '
          f'0x{smc_key[3]:02x} '
          f'0x{smc_key[4]:08x} '
          f'{bytetohex(smc_data)}')
    return


def gethdr(vmx, offset):
    # Read header into struct
    vmx.seek(offset)
    hdr = struct.unpack(HDR_PACK, vmx.read(HDR_LENGTH))
    vmx.seek(offset)
    return hdr


def getkey(vmx, offset):
    # Read key into struct
    vmx.seek(offset)
    smc_key = struct.unpack(KEY_PACK, vmx.read(KEY_LENGTH))
    vmx.seek(offset)
    return smc_key


def getdata(vmx, offset, smc_key):
    # Read data for key
    vmx.seek(offset + KEY_LENGTH)
    smc_data = vmx.read(smc_key[1])
    vmx.seek(offset)
    return smc_data


def dumpkeys(vmx, offset, count):

    print('Offset     Name Len Type Flag FuncPtr    Data')
    print('-------    ---- --- ---- ---- -------    ----')

    for i in range(count):

        # Read key into struct str and data byte str
        smc_key = getkey(vmx, offset)
        smc_data = getdata(vmx, offset, smc_key)

        # Dump entry
        printkey(offset, smc_key, smc_data)
        offset = offset + ROW_LENGTH

    return


def dumpsmc(name):

    with open(name, 'r+b') as f:

        # Memory map file
        vmx = mmap.mmap(f.fileno(), 0)

        print('File: ' + name)

        # Find the vSMC headers
        smc0_header = vmx.find(SMC_HEADER_V0) - 8
        smc1_header = vmx.find(SMC_HEADER_V1) - 8

        # Find '#KEY' keys
        smc0_key = vmx.find(KEY_KEY, smc0_header)
        smc1_key = vmx.find(KEY_KEY, smc1_header)

        # Dump first vSMC table
        print('\nappleSMCTableV0 (smc.version = "0")')
        hdr = gethdr(vmx, smc0_header)
        printhdr(smc0_header, hdr)
        dumpkeys(vmx, smc0_key, hdr[1])

        # Dump second vSMC table
        print('\nappleSMCTableV1 (smc.version = "1")')
        hdr = gethdr(vmx, smc1_header)
        printhdr(smc1_header, hdr)
        dumpkeys(vmx, smc1_key, hdr[1])

        # Tidy up
        vmx.close()
        f.close()

    return


def main():
    print('dumpsmc')
    print('-------')

    if len(sys.argv) >= 2:
        vmx_path = sys.argv[1]
    else:
        print('Please pass file name!')
        return

    if os.path.exists(vmx_path):
        dumpsmc(vmx_path)
    else:
        print('Cannot find file ' + vmx_path)
    return


if __name__ == '__main__':
    main()


esx/unlocker.sh
#!/bin/ash
set -x

# Absolute path to this script
SCRIPT=$(readlink -f "$0")
echo "$SCRIPT"

# Absolute folder to this script
SCRIPTPATH=$(dirname "$SCRIPT")
echo "$SCRIPTPATH"

# Change to script folder for processing
cd "$SCRIPTPATH"

# Get the GZIP/XZ/VMTAR'd VisorFS file from main bootbank (BOOTBANK1)
cp -fv /bootbank/vmx.v00 vmx.gz

# Extract XZ file from GZIP file
pigz -d -k vmx.gz
mv -fv vmx vmx.xz

# Extract VMTAR file from XZ file
xz -d --single-stream < vmx.xz > vmx.vtar

# Extract TAR file from VMTAR file
vmtar -x vmx.vtar -o vmx.tar

# Create a tmp folder for processing the TAR disk contents and extract files
mkdir -p ./tmp
tar xvf vmx.tar -C ./tmp

# Cleanup copied/extracted files
rm -fv vmx*

# Do the unlocking stuff
cd ./tmp

#  Compress files to TAR file
tar cvf ../vmx.tar *
cd ..

# Create VMTAR from TAR file
vmtar -c vmx.tar -v -o vmx.vtar

# Compress VMTAR into XZ file and add signature
xz --compress --stdout --lzma2=dict=2048KiB --check=crc32 --threads=8 vmx.vtar > vmx.xz
cat /usr/share/weasel/s.sigblob >> vmx.xz

# Compress XZ-sig to GZIP file
pigz -9 -p 60 -n -T -k -S .v00 vmx.xz
mv -fv vmx.xz.v00 vmx.v00

# Copy to main bootbank (BOOTBANK1)
#cp vmx.v00 /bootbank/vmx.v00


esxi-build-vib.py
#!/usr/bin/env python
"""
The MIT License (MIT)

Copyright (c) 2014-2021 Dave Parsons & Sam Bingner

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the 'Software'), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
"""

# Note to self:
# This is crap code used in place of a shell script
# GNU tar and ar must be used and not BSD version or error loading the tardisk in ESXi:
# brew install gnu-tar binutils

import datetime
from string import Template
import subprocess
import sys


if sys.version_info < (3, 6):
    sys.stderr.write('You need Python 3.6 or later\n')
    sys.exit(1)

# TODO: Change for a new release
VERSION = '3.0.1'
FILEVER = '301'
FILENAME = 'build/esxi-unlocker-301.tgz'

# Constants used in creating the VIB
NOW = datetime.datetime.now()
FILETIMESTAMP = f'{NOW:%Y%m%d%H%M.%S}'
VIBTIMESTAMP = (NOW.strftime('%Y-%m-%dT%H:%M:%S.%f+00:00'))
TOUCH = 'touch -t ' + FILETIMESTAMP
UNLOCKERTAR = '/usr/local/bin/gtar cvf build/unlocker.tar etc'
UNLOCKERTGZ = 'gzip -S .tgz -k build/unlocker.tar'
UNLOCKERMV = 'mv -vf build/unlocker.tar.tgz build/unlocker.tgz'
UNLOCKERVIB = 'agar qDv esxi-unlocker-301.vib descriptor.xml sig.pkcs7 unlocker.tgz'

# Template for the descriptor.xml file
DESCRIPTOR = Template(
    """<vib version="5.0">
	<type>bootbank</type>
	<name>esxi-unlocker</name>
	<version>$version</version>
	<vendor>unlocker</vendor>
	<summary>macOS Enabler for ESXi</summary>
	<description>macOS Enabler for ESXi</description>
	<release-date>$timestamp</release-date>
	<urls>
		<url key="kb">https://github.com/DrDonk/esxi-unlocker</url>
	</urls>
	<relationships>
		<depends></depends>
		<conflicts/>
		<replaces/>
		<provides/>
		<compatibleWith/>
	</relationships>
	<software-tags></software-tags>
	<system-requires>
		<maintenance-mode>true</maintenance-mode>
	</system-requires>
	<file-list>
		<file>etc/rc.local.d/unlocker.py</file>
	</file-list>
	<acceptance-level>community</acceptance-level>
	<live-install-allowed>false</live-install-allowed>
	<live-remove-allowed>false</live-remove-allowed>
	<cimom-restart>false</cimom-restart>
	<stateless-ready>true</stateless-ready>
	<overlay>false</overlay>
	<payloads>
		<payload name="unlocker.tgz" type="tgz" size="4105">
			<checksum checksum-type="sha-256">$sha256sum</checksum>
			<checksum checksum-type="sha-1" verify-process="gunzip">$sha1sum</checksum>
		</payload>
	</payloads>
</vib>
""")

def main():

    # Timestamp files for release
    print('\nTimestamping files...')
    subprocess.call(TOUCH + ' etc', shell=True)
    subprocess.call(TOUCH + ' etc/rc.local.d', shell=True)
    subprocess.call(TOUCH + ' etc/rc.local.d/unlocker.py', shell=True)

    # Build the gzipped tar file unlocker.tgz
    print('\nCreating unlocker.tgz...')
    subprocess.call(UNLOCKERTAR, shell=True)
    subprocess.call(TOUCH + ' unlocker.tar', shell=True)
    subprocess.call(UNLOCKERTGZ, shell=True)
    subprocess.call(UNLOCKERMV, shell=True)
    subprocess.call(TOUCH + ' unlocker.tgz', shell=True)

    # Get the sha1 and sha256 checksums needed for the VIB descriptor.xml
    sha1sum = subprocess.check_output('sha1sum build/unlocker.tar | cut -d " " -f 1', shell=True).decode().strip('\n')
    sha256sum = subprocess.check_output('sha256sum build/unlocker.tgz | cut -d " " -f 1', shell=True).decode().strip('\n')

    # Create a blank signing file
    subprocess.call(TOUCH + ' build/sig.pkcs7', shell=True)

    # Create the descriptor.xml filke
    d = dict()
    d['version'] = '3.0.1'
    d['timestamp'] = VIBTIMESTAMP
    d['sha256sum'] = sha256sum
    d['sha1sum'] = sha1sum

    descriptor_xml = DESCRIPTOR.substitute(d)
    with open('build/descriptor.xml', 'w') as f:
        f.write(descriptor_xml)
    subprocess.call(TOUCH + ' build/descriptor.xml', shell=True)

    # '/usr/local/bin/gtar czvf unlocker.tgz etc'
    # Build the distribution file esxi-unlocker-VER.tgz
    # print('\nCreating ' + FILENAME + '...')
    # subprocess.call(UNLOCKERVIB, shell=True)
    # subprocess.call(TOUCH + ' ' + FILENAME, shell=True)
    return

if __name__ == '__main__':
    if sys.platform == 'darwin':
        print('ESXi-Build for macOS')
        main()
    else:
        print('ESXi-Build only supported on macOS')


esxi-build.tgz.py
#!/usr/bin/env python
"""
The MIT License (MIT)

Copyright (c) 2014-2018 Dave Parsons & Sam Bingner

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the 'Software'), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
"""

# Note to self:
# This is crap code used in place of a shell script
# GNU tar and ar must be used and not BSD version or error loading the tardisk in ESXi:
# brew install gnu-tar binutils

import datetime
import subprocess
import sys


if sys.version_info < (3, 6):
    sys.stderr.write('You need Python 3.6 or later\n')
    sys.exit(1)

# TODO: Change for a new release
VERSION = '3.0.1'
FILEVER = '301'
FILENAME = 'build/esxi-unlocker-300.tgz'

NOW = datetime.datetime.now()
TIMESTAMP = f'{NOW:%Y%m%d%H%M.%S}'
TOUCH = 'touch -t ' + TIMESTAMP
GTARUNLOCKER = '/usr/local/bin/gtar czvf build/unlocker.tgz etc'
GTARDISTRIB = '/usr/local/bin/gtar czvf ' + FILENAME + \
              ' build/unlocker.tgz esxi-install.sh esxi-uninstall.sh esxi-smctest.sh readme.txt'


def main():

    # Timestamp files for release
    print('\nTimestamping files...')
    subprocess.call(TOUCH + ' readme.txt', shell=True)
    subprocess.call(TOUCH + ' esxi-install.sh', shell=True)
    subprocess.call(TOUCH + ' esxi-uninstall.sh', shell=True)
    subprocess.call(TOUCH + ' esxi-smctest.sh', shell=True)
    subprocess.call(TOUCH + ' etc', shell=True)
    subprocess.call(TOUCH + ' etc/rc.local.d', shell=True)
    subprocess.call(TOUCH + ' etc/rc.local.d/unlocker.py', shell=True)

    # Build the gzipped tar file unlocker.tgz
    print('\nCreating unlocker.tgz...')
    subprocess.call(GTARUNLOCKER, shell=True)
    subprocess.call(TOUCH + ' unlocker.tgz', shell=True)

    # Build the distribution file esxi-unlocker-VER.tgz
    print('\nCreating ' + FILENAME + '...')
    subprocess.call(GTARDISTRIB, shell=True)
    subprocess.call(TOUCH + ' ' + FILENAME, shell=True)


if __name__ == '__main__':
    if sys.platform == 'darwin':
        print('ESXi-Build for macOS')
        main()
    else:
        print('ESXi-Build only supported on macOS')


esxi-config-sandbox.py
#!/usr/bin/env python
"""
This is a simple method to modify the hostd XML file
Not using XML on ESXi Python as it does not preserve
formatting or comments.

(This could be sed but cannot find a suitable regex.)

"""
import sys


def testline(line, test):
    sline = line.lstrip()
    if sline == test:
        return True
    else:
        return False


def main():
    vmsvc = '<vmsvc>\n'
    useVMXSandbox = '<useVmxSandbox>false</useVmxSandbox>\n'
    enforceVmxSandbox = '<enforceVmxSandbox>false</enforceVmxSandbox>\n'

    with open('/etc/vmware/hostd/config.xml', 'r+') as f:
        data = f.readlines()

        # Search for the relevant XML tags
        i = 0
        vmsvcindex = 0
        sandboxindex = 0
        for line in data:

            if testline(line, vmsvc):
                vmsvcindex = i

            if testline(line, useVMXSandbox):
                sandboxindex = i

            # print(line, end='')
            i += 1

        # Simple toggle on or off depending if found
        if sandboxindex != 0 and sys.argv[1] == 'off':
            print('Removing useVmxSandbox')
            del data[sandboxindex]
        elif sandboxindex == 0 and sys.argv[1] == 'on':
            print('Adding useVmxSandbox')
            pad = len(data[vmsvcindex + 1]) - len(data[vmsvcindex + 1].lstrip())
            data.insert(vmsvcindex + 1, (" " * pad) + useVMXSandbox)
        else:
            pass

        # Rewrite the config.xml file
        f.seek(0)
        f.write(''.join(data))
        f.truncate()
        f.close()


if __name__ == '__main__':
    main()


esx/patchsmc.py
#!/usr/bin/env python3
# coding=utf-8

# SPDX-FileCopyrightText: © 2014-2022 David Parsons & Sam Bingner
# SPDX-License-Identifier: MIT

"""
vSMC Header Structure
=====================
Offset  Length  Struct Type Description
----------------------------------------
0x00/00 0x08/08 Q      ptr  Offset to key table
0x08/08 0x04/4  I      int  Number of private keys
0x0C/12 0x04/4  I      int  Number of public keys

vSMC Key Data Structure
Offset  Length  Struct Type Description
----------------------------------------
0x00/00 0x04/04 4s     int  Key name (byte reversed e.g. #KEY is YEK#)
0x04/04 0x01/01 B      byte Length of returned data
0x05/05 0x04/04 4s     int  Data type (byte reversed e.g. ui32 is 23iu)
0x09/09 0x01/01 B      byte Flag R/W
0x0A/10 0x06/06 6x     byte Padding
0x10/16 0x08/08 Q      ptr  Internal VMware routine
0x18/24 0x30/48 48B    byte Data

The internal VMware routines point to 4 variants:
AppleSMCHandleDefault
AppleSMCHandleNTOK
AppleSMCHandleNumKeys
AppleSMCHandleOSK
"""

import codecs
import mmap
import os
import struct
import sys


# Constants for header and key access
HDR_PACK = '=QII'
HDR_LENGTH = 16
KEY_PACK = '=4sB4sB6xQ'
KEY_LENGTH = 24

# Setup hex string for vSMC headers
# These are the private and public key counts
SMC_HEADER_V0 = b'\xF2\x00\x00\x00\xF0\x00\x00\x00'
SMC_HEADER_V1 = b'\xB4\x01\x00\x00\xB0\x01\x00\x00'

# Keys we use
KEY_KEY = b'YEK#'
LKS_KEY = b'SKL+'
OSK0_KEY = b'0KSO'
OSK1_KEY = b'1KSO'
KPST_KEY = b'TSPK'
KPPW_KEY = b'WPPK'

# Haiku
OSK0 = codecs.encode('bheuneqjbexolgurfrjbeqfthneqrqcy', 'rot_13').encode('UTF-8')
OSK1 = codecs.encode('rnfrqbagfgrny(p)NccyrPbzchgreVap', 'rot_13').encode('UTF-8')

# Hogwarts
KPPW = codecs.encode('FcrpvnyvfEriryvb', 'rot_13').encode('UTF-8')

# ELF Magic
ELF_MAGIC = b'\x7fELF'

if sys.version_info < (3, 6):
    sys.stderr.write('You need Python 3.6 or later\n')
    sys.exit(1)


def bytetohex(data):
    return ''.join('{:02X} '.format(c) for c in data)


def printhdr(offset, hdr):
    print(f'File Offset : 0x{offset:08x}')
    print(f'Keys Offset : 0x{hdr[0]:08x}')
    print(f'Private Keys: 0x{hdr[1]:04x}/{hdr[1]}')
    print(f'Public Keys : 0x{hdr[2]:04x}/{hdr[2]}')
    return


def printkey(offset, smc_key, smc_data):
    # Format smc_type as cannot use \ in f-strings
    smc_type = smc_key[2][::-1].replace(b'\x00', b' ').decode('UTF-8')
    print(f'0x{offset:08x} '
          f'{smc_key[0][::-1].decode("UTF-8")} '
          f'{smc_key[1]:03d} '
          f'{smc_type} '
          f'0x{smc_key[3]:02x} '
          f'0x{smc_key[4]:08x} '
          f'{bytetohex(smc_data)}')
    return


def gethdr(vmx, offset):
    # Read header into struct
    vmx.seek(offset)
    hdr = struct.unpack(HDR_PACK, vmx.read(HDR_LENGTH))
    vmx.seek(offset)
    return hdr


def getkey(vmx, offset):
    # Read key into struct
    vmx.seek(offset)
    smc_key = struct.unpack(KEY_PACK, vmx.read(KEY_LENGTH))
    vmx.seek(offset)
    return smc_key


def setkey(vmx, offset, smc_key):
    # Write key from struct
    vmx.seek(offset)
    vmx.write(struct.pack(KEY_PACK, smc_key[0], smc_key[1], smc_key[2], smc_key[3], smc_key[4]))
    vmx.flush()
    vmx.seek(offset)
    return


def getdata(vmx, offset, smc_key):
    # Read data for key
    vmx.seek(offset + KEY_LENGTH)
    smc_data = vmx.read(smc_key[1])
    vmx.seek(offset)
    return smc_data


def setdata(vmx, offset, smc_data):
    # Write data for key
    vmx.seek(offset + KEY_LENGTH)
    vmx.write(smc_data)
    vmx.flush()
    vmx.seek(offset)
    return


def patchkppw(vmx, offset, data):
    # Get the KPPW key and data
    smc_key = getkey(vmx, offset)
    smc_data = getdata(vmx, offset, smc_key)
    key = smc_key[0][::-1].decode('UTF-8')
    print(f'{key} Key Before:')
    printkey(offset, smc_key, smc_data)

    # Set the data value
    setdata(vmx, offset, data)

    # Get patched KPPW key and data
    smc_key = getkey(vmx, offset)
    smc_data = getdata(vmx, offset, smc_key)
    print(f'{key} Key After:')
    printkey(offset, smc_key, smc_data)

    return


def patchosk(vmx, offset, ptr, data):
    # Get the OSK key and data
    smc_key = getkey(vmx, offset)
    smc_data = getdata(vmx, offset, smc_key)
    key = smc_key[0][::-1].decode('UTF-8')
    smc_osk_ptr = smc_key[4]
    print(f'{key} Key Before:')
    printkey(offset, smc_key, smc_data)

    # AppleSMCHandleOSK replaced with AppleSMCHandleDefault
    temp = list(smc_key)
    temp[4] = ptr
    smc_key = tuple(temp)
    setkey(vmx, offset, smc_key)

    # Set the data value
    setdata(vmx, offset, data)

    # Get patched OSK key and data
    smc_key = getkey(vmx, offset)
    smc_data = getdata(vmx, offset, smc_key)
    print(f'{key} Key After:')
    printkey(offset, smc_key, smc_data)

    return smc_osk_ptr


def patchsmc(name):
    with open(name, 'r+b') as f:
        print('File: ' + name + '\n')

        # Memory map file
        vmx = mmap.mmap(f.fileno(), 0)

        # Find the vSMC headers
        smc0_header = vmx.find(SMC_HEADER_V0) - 8
        smc1_header = vmx.find(SMC_HEADER_V1) - 8

        # Find '#KEY' keys
        smc0_key = vmx.find(KEY_KEY)
        smc1_key = vmx.rfind(KEY_KEY)

        # Find '+LKS' key
        smc0_lks = vmx.find(LKS_KEY, smc0_key)
        smc1_lks = vmx.find(LKS_KEY, smc1_key)

        # Find 'OSK0' keys
        smc0_osk0 = vmx.find(OSK0_KEY, smc0_key)
        smc1_osk0 = vmx.find(OSK0_KEY, smc1_key)

        # Find 'OSK1' keys
        smc0_osk1 = vmx.find(OSK1_KEY, smc0_key)
        smc1_osk1 = vmx.find(OSK1_KEY, smc1_key)

        # Find 'KPST' key
        smc1_kpst = vmx.find(KPPW_KEY, smc1_key)

        # Find 'KPPW' key
        smc1_kppw = vmx.find(KPPW_KEY, smc1_key)

        # Check to see if we have already patched the vSMC in the file
        osk0 = getdata(vmx, smc1_osk0, getkey(vmx, smc1_osk0))
        osk1 = getdata(vmx, smc1_osk1, getkey(vmx, smc1_osk1))
        kppw = getdata(vmx, smc1_kppw, getkey(vmx, smc1_kppw))

        if osk0 == OSK0 and osk1 == OSK1 and kppw == KPPW:
            print(f'File {name} is already patched')
            return

        # Patch first vSMC table
        print('\nappleSMCTableV0 (smc.version = "0")')
        hdr = gethdr(vmx, smc0_header)
        printhdr(smc0_header, hdr)

        # Get the +LKS key data routine for OSK0/1
        smc_key = getkey(vmx, smc0_lks)
        smc_default_ptr = smc_key[4]

        # Patch OSK0 key
        patchosk(vmx, smc0_osk0, smc_default_ptr, OSK0)

        # Patch OSK1 key
        patchosk(vmx, smc0_osk1, smc_default_ptr, OSK1)

        # Patch second vSMC table
        print('\nappleSMCTableV1 (smc.version = "1")')
        hdr = gethdr(vmx, smc1_header)
        printhdr(smc1_header, hdr)

        # Get the +LKS key data routine for OSK0/1
        smc_key = getkey(vmx, smc1_lks)
        smc_default_ptr = smc_key[4]

        # Patch OSK0 key
        patchosk(vmx, smc1_osk0, smc_default_ptr, OSK0)

        # Patch OSK1 key & get the output for ELF patching
        smc_osk_ptr = patchosk(vmx, smc1_osk1, smc_default_ptr, OSK1)

        # Patch KPPW key as a marker
        patchkppw(vmx, smc1_kppw, KPPW)

        # Patch relocation records if ELF executable
        vmx.seek(0)
        magic = vmx.read(4)
        if magic == ELF_MAGIC:
            # Find matching RELA record in .rela.dyn in ELF files
            print(f'\nModifying ELF RELA records from 0x{smc_osk_ptr:08x} -> 0x{smc_default_ptr:08x}')

            # Start at offset 0 and work thru file counting patches
            i = 0
            j = 0

            # Repack ints to bytes for find
            packed_old_ptr = struct.pack('=Q', smc_osk_ptr)
            packed_new_ptr = struct.pack('=Q', smc_default_ptr)

            # Loop while lower than SMC V0 #KEY offset
            while True:
                i = vmx.find(packed_old_ptr, i)
                if i != -1:
                    print(f'Relocation modified at: 0x{i:08x}')
                    vmx.seek(i)
                    vmx.write(packed_new_ptr)
                    i += 1
                    j += 1
                else:
                    break

            # There should be 4 relocation records that need patching
            if j != 4:
                print(f'ELF file RELA error count: {j}')

        # Tidy up
        vmx.flush()
        f.flush()
        vmx.close()
        f.close()


def ispatched(name):
    with open(name, 'r+b') as f:
        print('File: ' + name + '\n')

        # Memory map file
        vmx = mmap.mmap(f.fileno(), 0)

        # Is SpecialisRevelio present?
        if vmx.find(KPPW) == -1:
            flag = True
        else:
            flag = False

    # Tidy up
    vmx.close()
    f.close()
    return flag


def main():
    print('patchsmc')
    print('--------')

    if len(sys.argv) >= 2:
        filename = sys.argv[1]
    else:
        print('Please pass file name!')
        return

    if os.path.isfile(filename):
        patchsmc(filename)
    else:
        print('Cannot find file ' + filename)
    return


if __name__ == '__main__':
    main()
